aspnetcore/release-notes/aspnetcore-3.0.md

```diff
@@ -134,7 +134,184 @@ For more information, see <xref:blazor/index>.
+
+To make providing these cues easier, we’ve expanded the SignalR client API to include `onreconnecting` and `onreconnected` event handlers. The first of these handlers, `onreconnecting`, gives developers a good opportunity to disable UI or to let users know the app is offline.
+
+```javascript
+connection.onreconnecting((error) => {
+    const status = `Connection lost due to error "${error}". Reconnecting.`;
+    document.getElementById("messageInput").disabled = true;
+    document.getElementById("sendButton").disabled = true;
+    document.getElementById("connectionStatus").innerText = status;
+});
+```
+
+Likewise, the `onreconnected` handler gives developers an opportunity to update the UI once the connection is reestablished.
+
+```javascript
+connection.onreconnected((connectionId) => {
+    const status = `Connection reestablished. Connected.`;
+    document.getElementById("messageInput").disabled = false;
+    document.getElementById("sendButton").disabled = false;
+    document.getElementById("connectionStatus").innerText = status;
+});
+```
+
+SignalR now provides a custom resource to authorization handlers when a hub method requires authorization. The resource is an instance of `HubInvocationContext`. The `HubInvocationContext` includes the `HubCallerContext`, the name of the hub method being invoked, and the arguments to the hub method.
+
+Consider the example of a chat room allowing multiple organization sign-in via Azure Active Directory. Anyone with a Microsoft account can sign in to chat, but only members of the owning organization should be able to ban users or view users’ chat histories. Furthermore, we might want to restrict certain functionality from certain users. Using the updated features in Preview 7, this is entirely possible. Note how the `DomainRestrictedRequirement` serves as a custom `IAuthorizationRequirement`. Now that the `HubInvocationContext` resource parameter is being passed in, the internal logic can inspect the context in which the Hub is being called and make decisions on allowing the user to execute individual Hub methods.
+
+```csharp
+public class DomainRestrictedRequirement :
+    AuthorizationHandler<DomainRestrictedRequirement, HubInvocationContext>,
+    IAuthorizationRequirement
+{
+    protected override Task HandleRequirementAsync(AuthorizationHandlerContext context,
+        DomainRestrictedRequirement requirement,
+        HubInvocationContext resource)
+    {
+        if (IsUserAllowedToDoThis(resource.HubMethodName, context.User.Identity.Name) &&
+            context.User != null &&
+            context.User.Identity != null &&
+            context.User.Identity.Name.EndsWith("@jabbr.net", StringComparison.OrdinalIgnoreCase))
```

Comment:
All this null checking seems unnecessary since `context.User.Identity.Name` is already passed to IsUserAllowedToDoThis above. IsUserAllowedToDoThis should verify the name ends with "@jabbr.net" regardless.

BTW

```c#
            context.User != null &&
            context.User.Identity != null &&
            context.User.Identity.Name.EndsWith("@jabbr.net", StringComparison.OrdinalIgnoreCase)
```

and 

```c#
            context?.User?.Identity?.Name?.EndsWith("@jabbr.net", StringComparison.OrdinalIgnoreCase) == true
```

are equivalent expressions, but the latter is probably a little too cute for a doc sample.

Reply:
The best approach is probably to add exit early if no user name is set. Ex:

```
if (context.User?.Identity?.Name == null)
{
    return Task.CompletedTask;
}
```

Then you can remove the other null checking.

Reply:
@halter73 I'll stop working on #14371 if you can move over to #14371. That way I can accept your suggestion commits without copy/paste.